### ISSUE TITLE
Decimal separator conversion issue (e.g., "." to "," and vice versa)

### ISSUE THREAD

**Issue by deasydoesit** [helium/hotspot-app#621]
App version 3.2.2 - iOS
Locale = USA / EU

Steps to Reproduce :

- On an app with locale set to US, scan a `payment` QR code with an `amount` using the European decimal separator, `,` (e.g., `312,0919210`).
- The `amount` shown in the payment staging area ignores the decimal separator and instead displays the `amount` as a whole number (e.g., `3,120,919,210`)

- Similarly, on an app with locale set to EU, scan a `payment` QR code with an `amount` using the American decimal separator, `.` (e.g., `312.0919210`).
- The `amount` shown in the payment staging area ignores the decimal separator and instead displays the `amount` as a whole number (e.g., `3,120,919,210`)

Expected Result :

- Respect the decimal separator across regions.

The issue is likely occurring in the `parseAmount` function in `src/utils/transactions.ts`

**Comment by pe1mew** [comment]
Currently, only HNT can be transferred as whole numbers (int) without fractions, when using a regional setting where the decimal separator is set to a comma ",". For example, this is the case in the Netherlands (NL).

The button ",-" is entered and then removed when entering a fraction.

The problem is also observed when transferring HNT to Helium Rising using their Qr-code.

**Comment by PaulVMo** [comment]
Also, for manually entered transactions, it looks like there are some issues. Any transaction entered that is more than 1000 in US notation (e.g. 1,000) have the digits after the comma dropped when the transaction is submitted to the block chain. So what looks like a 1,000 HNT transaction is processed as 1 HNT.

**Comment by cokes518** [comment]
what do you use to generate the payment QR code?

**Comment by cokes518** [comment]
and what happens if you generate a EU format QR code and scan with an EU Locale app?

**Comment by cokes518** [comment]
> Currently, only HNT can be transferred as whole numbers (int) without fractions, when using a regional setting where the decimal separator is set to a comma ",". For example, this is the case in the Netherlands (NL).
> 
> The button ",-" is entered and then removed when entering a fraction.
> 
> The problem is also observed when transferring HNT to Helium Rising using their Qr-code.

can you please file this as a separate issue?

**Comment by pe1mew** [comment]
See: https://github.com/helium/hotspot-app/issues/621#issuecomment-852276789

**Comment by cokes518** [comment]
> App version 3.2.2 - iOS
> Locale = USA / EU
> 
> Steps to Reproduce :
> 
> * On an app with locale set to US, scan a `payment` QR code with an `amount` using the European decimal separator, `,` (e.g., `312,0919210`).
> * The `amount` shown in the payment staging area ignores the decimal separator and instead displays the `amount` as a whole number (e.g., `3,120,919,210`)
> * Similarly, on an app with locale set to EU, scan a `payment` QR code with an `amount` using the American decimal separator, `.` (e.g., `312.0919210`).
> * The `amount` shown in the payment staging area ignores the decimal separator and instead displays the `amount` as a whole number (e.g., `3,120,919,210`)
> 
> Expected Result :
> 
> * Respect the decimal separator across regions.
> 
> The issue is likely occurring in the `parseAmount` function in `src/utils/transactions.ts`

This doesn't look like a new issue. The main problem comes from the QR code is generated using a number format mismatched with the device (QR sends commas, app expects periods) and the device just gives up, which is a poor experience.

One thing we can do, is have the QR code include the locale so the app will know what to display when it expects a number with commas and periods, and whether to treat them as group separators or decimal separators.

A workaround here would be to generate QRs that match the regional settings of the user scanning the QR code.

**Comment by cokes518** [comment]
> See: [#621 (comment)](https://github.com/helium/hotspot-app/issues/621#issuecomment-852276789)

Not sure why you referenced that comment, but that is also completely separate issue and is fixed, waiting to be merged.

**Comment by tyler-whitman** [comment]
As of now the app expects the QR data to be in US number format. I see two solutions to the problem.

1. We continue to force QR data to be in US format (the app will still display the number in the devices native locale, just the data in the QR needs to be US format).
2. We add a locale property to the QR data so that we know format the data is in.

Either way the person generating the QR will need to update, as the app wont be able to succinctly determine the locale format of the number in the QR data automatically. I would prefer keeping the QR data in US format and not adding a new property to the QR data just to keep it smaller and less complicated code wise.

**Comment by cokes518** [comment]
> We continue to force QR data to be in US format (the app will still display the number in the devices native locale, just the data in the QR needs to be US format).

as painful as it is to be non-i18n friendly, this is the best path forward. Because the iOS and Android app environments treat language and locale/region differently, it's hard to catch all the possible cases. By restricting QR codes to be US-only (commas as group separators and periods as decimal separators), the app will always know how to parse the data, and display it to the user's desired settings.

**Comment by deasydoesit** [comment]
Mandating QR code formation with the `value` in US format (e.g., with a `.` for the decimal separator) is a fine solution. 

As an effort to preempt any issues with non-US `value` formatted QR creators, perhaps documentation could be included / updated to clarify such a requirement?

**Comment by cokes518** [comment]
> As an effort to preempt any issues with non-US value formatted QR creators, perhaps documentation could be included / updated to clarify such a requirement?

the wallet CLI current uses the JSON spec to represent numbers with no commas and uses periods as decimal separators. any other QR codes generated likely used a JSON serializer out of JSON spec, so we cannot reasonably enforce everyone to do this.

**Comment by deasydoesit** [comment]
I'm not suggesting to attempt any QR code `value` standardization enforcement, but rather to highlight the requirement in the `hotspot-app` documentation.

**Comment by tyler-whitman** [comment]
There is also still one bug to fix here in the app. This case should work, so ill fix it and get that into the next release.

```
Similarly, on an app with locale set to EU, scan a payment QR code with an amount using the American decimal separator, . (e.g., 312.0919210).

The amount shown in the payment staging area ignores the decimal separator and instead displays the amount as a whole number (e.g., 3,120,919,210)
```

**Comment by cokes518** [comment]
fixed in 3.2.3
